### PR TITLE
Implement FRI committed layers

### PR DESCRIPTION
**Pull request by varunthakore**
Implements FRI layers and committed layers corresponding to the following Stone library:

https://github.com/starkware-libs/stone-prover/tree/main/src/starkware/fri

This PR has dependency on #18.

### CODE REVIEW

File: fri/src/layers.rs [hunk 1]
```diff
@@ -0,0 +1,378 @@
...
+    }
+
+    fn eval_at_points(&self, required_indices: &[usize]) -> Vec<F> {
+        required_indices
+            .iter()
+            .map(|&i| self.evaluation[i])
+            .collect()
+    }
+
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>> {
```
Comment (rot256) [inline fri/src/layers.rs:62]: Why is this method needed? Can't the user just wrap the thing in an Arc himself?

This also copies the underlying struct, I would imagine that the goal was to have a number of references to the *same* immutable instance?

File: fri/src/layers.rs [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+
+    fn eval_at_points(&self, _required_indices: &[usize]) -> Vec<F> {
+        unimplemented!("Should never be called")
+    }
+
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>> {
+        Arc::new(FriLayerProxy {
+            layer_size: self.layer_size,
+            domain: self.domain,
+            prev_layer: Arc::clone(&self.prev_layer),
```
Comment (rot256) [inline fri/src/layers.rs:122]: `self.prev_layer.clone()` does the same.

File: fri/src/layers.rs [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+    fn get_layer_size(&self) -> usize;
+    fn get_domain(&self) -> E;
+    fn get_layer(&self) -> Result<Vec<F>, Error>;
+    fn eval_at_points(&self, required_indices: &[usize]) -> Vec<F>;
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>>;
+}
+
+#[derive(Clone)]
+pub struct FriLayerReal<F: PrimeField, E: EvaluationDomain<F>> {
+    layer_size: usize,
```
Comment (rot256) [inline fri/src/layers.rs:18]: `layer_size` is a redundant field: it looks to be the same as `domain.size()`?

File: fri/src/layers.rs [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+impl<F: PrimeField, E: EvaluationDomain<F>> FriLayerReal<F, E> {
+    pub fn new(domain: E, evaluation: Vec<F>) -> Self {
+        Self {
+            layer_size: domain.size(),
+            domain,
+            evaluation,
+        }
+    }
+
+    pub fn new_from_prev_layer(prev_layer: &dyn FriLayer<F, E>) -> Self {
```
Comment (rot256) [inline fri/src/layers.rs:33]: This just clones the layer?

**Review by rot256 — APPROVED** [review]
A couple questions, otherwise LGTM.

File: fri/src/layers.rs (hunk 1 above)
Comment (varunthakore) [reply fri/src/layers.rs:62]: Yes, removed this function.

File: fri/src/layers.rs (hunk 2 above)
Comment (varunthakore) [reply fri/src/layers.rs:122]: Removed the `clone_arc` function.

File: fri/src/layers.rs (hunk 3 above)
Comment (varunthakore) [reply fri/src/layers.rs:18]: Removed `layer_size`.

File: fri/src/layers.rs (hunk 4 above)
Comment (varunthakore) [reply fri/src/layers.rs:33]: It is used to create `FriLayerReal` from the previous `FriLayerProxy`, both the structs implement `FriLayer` trait.

**Comment by varunthakore** [comment]
Closing this PR since #24 is merged.

**Pull request closed without merging**